### PR TITLE
IBX-607: Added method 'first' to ItemListInterface

### PR DESCRIPTION
**Pull request by ciastektk**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-607](https://issues.ibexa.co/browse/IBX-607)
| **Type**                                   | improvement
| **Target Ibexa DXP version** | `v4.0`
| **BC breaks**                          | no
| **Doc needed**                       | no

This PR provides added `first` method to `ItemTypeInterface` to get first item from the item list. There is also changed ItemNotFoundException passing `itemId` and `language` arguments are optional now. 

#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [ ] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/php-dev-team`).

### CODE REVIEW

**Review by alongosz — CHANGES_REQUESTED** [review]
Missing minimal test coverage (empty list, 1 element list, 2+ elements list)

File: src/lib/Value/Storage/ItemList.php [hunk 1]
```diff
@@ -80,6 +80,15 @@ public function slice(int $offset, ?int $length = null): ItemListInterface
         return new self(array_slice($this->items, $offset, $length));
     }
 
+    public function first(): ItemInterface
+    {
+        if (empty($this->items)) {
+            throw new ItemNotFoundException();
```
Comment (Steveb-p) [inline src/lib/Value/Storage/ItemList.php:86]: Should be `OutOfBoundsException` instead.

As discussed privately, I'm generally against using exceptions for cases like this, as empty collection is a valid state and something to be expected. What's more, we're forcing consumers to either handle the exception (performance penalty) or check if the collection is empty beforehand (possible performance penalty).

Considering the alternative - allow `null` to be returned and force consumers to check against `null` - in my opinion it's superior code-wise, performance-wise and more friendly to static analysis.

**Review by Steveb-p — CHANGES_REQUESTED** [review]

File: tests/lib/Value/ItemListTest.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+namespace Ibexa\Tests\Personalization\Value;
+
+use Ibexa\Contracts\Personalization\Value\ItemInterface;
+use Ibexa\Contracts\Personalization\Value\ItemListInterface;
+use Ibexa\Personalization\Value\Storage\ItemList;
+use Ibexa\Tests\Personalization\Creator\DataSourceTestItemCreator;
+use Ibexa\Tests\Personalization\Storage\AbstractDataSourceTestCase;
+use OutOfBoundsException;
+
+final class ItemListTest extends AbstractDataSourceTestCase
```
Comment (Steveb-p) [inline tests/lib/Value/ItemListTest.php:18]: Missing `@covers` annotation.

Comment (ciastektk) [reply tests/lib/Value/ItemListTest.php:21]: Added: https://github.com/ezsystems/ezrecommendation-client/pull/73/commits/87c8111b108a579c8a6de0bb629e8cf30772ba97

File: src/lib/Exception/ItemNotFoundException.php [hunk 3]
```diff
@@ -12,10 +12,16 @@
...
 final class ItemNotFoundException extends NotFoundException
 {
-    public function __construct(string $itemId, string $language, int $code = 0, Throwable $previous = null)
+    public function __construct(?string $itemId = null, ?string $language = null, int $code = 0, Throwable $previous = null)
     {
+        $message = 'Item not found';
+
+        if (isset($itemId, $language)) {
+            $message .= sprintf(' with id: %s and language: %s', $itemId, $language);
+        }
```
Comment (Steveb-p) [inline src/lib/Exception/ItemNotFoundException.php:21]: Since now you have `OutOfBoundsException`, these changes are no longer needed. Considering they introduced relaxed types, they should be reverted back :smile_cat:

File: tests/lib/Value/ItemListTest.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+ */
+final class ItemListTest extends AbstractDataSourceTestCase
+{
+    public function testThrowExceptionWhenFirstElementNotExists(): void
+    {
+        $this->expectException(OutOfBoundsException::class);
+        $this->expectExceptionMessage('Collection is empty');
+
+        $itemList = new ItemList([]);
+        $itemList->first();
```
Comment (alongosz) [inline tests/lib/Value/ItemListTest.php:29]: If exception for some reason would be thrown in the constructor, this test would stop testing what it should. Not that it will happen in this case, but the PHPUnit syntax was changed from annotation to method, so you can place exception expectation right before it is expected to happen:
```suggestion
        $itemList = new ItemList([]);
        
        $this->expectException(OutOfBoundsException::class);
        $this->expectExceptionMessage('Collection is empty');
        
        $itemList->first();
```

File: tests/lib/Value/ItemListTest.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        self::assertEquals($expectedItem, $itemList->first());
+    }
+
+    /**
+     * @phpstan-return iterable<array{
+     *  \Ibexa\Contracts\Personalization\Value\ItemListInterface,
+     *  \Ibexa\Contracts\Personalization\Value\ItemInterface
+     * }>
+     */
+    public function provideForTestFirst(): iterable
```
Comment (alongosz) [inline tests/lib/Value/ItemListTest.php:46]: grammar: `provideDataForTestFirst`

**Review by alongosz — CHANGES_REQUESTED** [review]

**Review by Steveb-p — APPROVED** [review]

**Review by alongosz — APPROVED** [review]
